### PR TITLE
[ML] Limit in flight requests when indexing model download parts

### PR DESCRIPTION
**Pull request by davidkyle**
#111684 improved the model install time by using multiple streams and threads to download and write the model parts. The change was reverted in #112961 after it was discovered the be the cause of Out Of Memory exceptions.

The design relied on using a fixed size thread pool to limit the concurrent downloads and hence also manage memory usage. However, the indexing of the downloaded document was performed async which meant a new download request would be forked and executed while the write request was still in flight leading to large numbers of in flight requests. The fix here is to block on the index write. 

The first commit is the revert of the revert, the later commits introduce the blocking write and reuse a byte buffer that was being recreated for every downloaded part.  Allocating that byte buffer is now protected by a circuit breaker. 

Labelled as a non issue because the code that caused the OOM was reverted before it made it to a production environment.

### CODE REVIEW

**Review by pxsalehi — COMMENTED** [review]
LGTM for the newer changes. Although, ideally there should be a test that shows the number of parallel downloads is limited. In any case, I'm gonna leave the final LGTM to ML, if that's ok.

File: x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java
Comment (prwhelan) [inline x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java:279]: This is the part that is blocking and/or slowing down the indexing, correct? We'll wait for the client response rather than continue asynchronously?

File: x-pack/plugin/ml-package-loader/src/test/java/org/elasticsearch/xpack/ml/packageloader/action/ModelLoaderUtilsTests.java
Comment (prwhelan) [inline x-pack/plugin/ml-package-loader/src/test/java/org/elasticsearch/xpack/ml/packageloader/action/ModelLoaderUtilsTests.java:134]: nice

**Review by prwhelan — APPROVED** [review]

**Comment by davidkyle** [comment]
@pxsalehi I pushed a small change to add a circuit breaker check before allocating the byte butter. Please can you take a look at https://github.com/elastic/elasticsearch/pull/112992/commits/78905e5f92984f48354b46d64d3aa27965c9fffc. Thanks

File: x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java
Comment (davidkyle) [reply x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java:295]: The gain is from using 5 threads to stream the download and index the parts. The non-blocking write meant that we had more than 5 in flight requests (the download is faster than the indexing) and that was causing the OOM. In order to limit the number of requests to at most 5 there has to be some element of blocking. Model download uses it's a dedicated thread pool so the block does not starve other parts of the code of resources

**Comment by davidkyle** [comment]
@elasticmachine update branch

**Review by pxsalehi — COMMENTED** [review]
> @pxsalehi I pushed a small change to add a circuit breaker check before allocating the byte butter. Please can you take a look at https://github.com/elastic/elasticsearch/commit/78905e5f92984f48354b46d64d3aa27965c9fffc. Thanks

LGTM. Thanks.

**Pull request merged**
